### PR TITLE
Prepend the attempt to keep the valid extension

### PR DESCRIPTION
**Pull request by Fokko**
Hi Gil,

I've patched some issues with writing files other than Parquet. We've discussed this earlier, but when writing json/csv, the compression is ignored upon reading. This is because the `attempt_` is appended and therefore the extension isn't picked up anymore.

As from http://comphadoop.weebly.com/

_If the input files are compressed, they will be decompressed automatically as they are read by MapReduce, using the filename extension to determine which codec to use. For example, a file ending in .gz can be identified as gzip-compressed file and thus read with GzipCodec._

I can imagine that this is a big change. If there are any further questions, please let me know. But I think it is important to get this in, to be compatible with csv/json/text/ etc.

I also did a test with the current master of stocator and my branch:

```
scala> val data = "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum."
data: String = Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.

scala> val arr = data.split(" ")
arr: Array[String] = Array(Lorem, Ipsum, is, simply, dummy, text, of, the, printing, and, typesetting, industry., Lorem, Ipsum, has, been, the, industrys, standard, dummy, text, ever, since, the, 1500s,, when, an, unknown, printer, took, a, galley, of, type, and, scrambled, it, to, make, a, type, specimen, book., It, has, survived, not, only, five, centuries,, but, also, the, leap, into, electronic, typesetting,, remaining, essentially, unchanged., It, was, popularised, in, the, 1960s, with, the, release, of, Letraset, sheets, containing, Lorem, Ipsum, passages,, and, more, recently, with, desktop, publishing, software, like, Aldus, PageMaker, including, versions, of, Lorem, Ipsum.)

scala> val distData = sc.parallelize(arr)
distData: org.apache.spark.rdd.RDD[String] = ParallelCollectionRDD[1] at parallelize at <console>:28

scala> val df = distData.toDF("word")
df: org.apache.spark.sql.DataFrame = [word: string]

scala> df.write.option("compression", "bzip2").json("swift2d://druid.mobpro/test-prepend/")

scala> spark.read.json("swift2d://druid.mobpro/test-prepend/").show()
+------------+                                                                  
|        word|
+------------+
|    survived|
|         not|
|        only|
|        five|
|  centuries,|
|         but|
|        also|
|         the|
|        leap|
|        into|
|  electronic|
|typesetting,|
|   remaining|
| essentially|
|  unchanged.|
|          It|
|         was|
| popularised|
|          in|
|         the|
+------------+
only showing top 20 rows
```

And stocator's master:
```
scala> val data = "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum."
data: String = Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.

scala> val arr = data.split(" ")
arr: Array[String] = Array(Lorem, Ipsum, is, simply, dummy, text, of, the, printing, and, typesetting, industry., Lorem, Ipsum, has, been, the, industrys, standard, dummy, text, ever, since, the, 1500s,, when, an, unknown, printer, took, a, galley, of, type, and, scrambled, it, to, make, a, type, specimen, book., It, has, survived, not, only, five, centuries,, but, also, the, leap, into, electronic, typesetting,, remaining, essentially, unchanged., It, was, popularised, in, the, 1960s, with, the, release, of, Letraset, sheets, containing, Lorem, Ipsum, passages,, and, more, recently, with, desktop, publishing, software, like, Aldus, PageMaker, including, versions, of, Lorem, Ipsum.)

scala> val distData = sc.parallelize(arr)
distData: org.apache.spark.rdd.RDD[String] = ParallelCollectionRDD[0] at parallelize at <console>:28

scala> val df = distData.toDF("word")
df: org.apache.spark.sql.DataFrame = [word: string]

scala> df.write.option("compression", "bzip2").json("swift2d://druid.mobpro/test-append/")

scala> spark.read.json("swift2d://druid.mobpro/test-append/").show()
+--------------------+
|     _corrupt_record|
+--------------------+
|BZh91AY&SY�{ƸK]...|
+--------------------+
```

The data on the SWIFT store using Swift explorer:

image

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by gilv** [comment]
@Fokko yes, it's important to have compression work with non Parquet as well. Not sure how popular this usage, but we need to have it supported. I also managed to reproduce it based on your input. I think the correct solution need to be a bit different, but i didn't tested it. Will your patch work if you keep existing name pattern "part-XYZ-attempt-id" but just add type at the end, like "part-XYZ-attempt-id.gzip" ?

**Comment by Fokko** [comment]
Hi Gil, yes that would also work. It might have less impact. Most of the time the extension is a compound, like `.json.gz`, or `.csv.bz2`, etc

**Comment by gilv** [comment]
@Fokko I guess only .gz or .bz2 is what is really relevant ( without .json. or .csv. )

**Comment by gilv** [comment]
@Fokko we need to have part-X-attempt-Y since it is important for list flows.. there parts have to be in certain order. Seems all what is needed to fix is just to add object compression type after the part-X-attempt-Y. This will do the trick. What do you think?

**Comment by gilv** [comment]
@Fokko are you also using COS connector ( for https://www.ibm.com/cloud/object-storage ) in addition to the Swift connector? It's not related this particular PR which is for Swift connector, but seems COS connector also has the same issue with compressed text files

**Comment by Fokko** [comment]
Hi @gilv. Thanks for looking into it. I'm only using SWIFT at the moment. 

> I guess only .gz or .bz2 is what is really relevant ( without .json. or .csv. )

Personally I would strongly argue to keep the whole extension, both `.csv.bz2` and not only the compression one. 

> we need to have part-X-attempt-Y since it is important for list flows.. there parts have to be in certain order. Seems all what is needed to fix is just to add object compression type after the part-X-attempt-Y.

I agree that we need to move to the `part-X-attempt-Y.format.compression` format. This change is far less intrusive.

> are you also using COS connector in addition to the Swift connector? It's not related this particular PR which is for Swift connector, but seems COS connector also has the same issue with compressed text files

I'm not using the COS storage, we're working on an OpenShift platform. But after making SWIFT compatible, it shouldn't be hard to fix the same issue for the COS connector.

If time allows today, I'll move to the `part-X-attempt-Y.format.compression` format and test it on my development environment.

Cheers, Fokko

**Comment by gilv** [comment]
@Fokko sure, let's keep the entire object type like .csv.bz2.. thanks

**Comment by Fokko** [comment]
image

**Comment by Fokko** [comment]
One of the problems is, that is might have problems reading files that are in the old format. But I still have to test this.

**Comment by gilv** [comment]
@Fokko let me know when you are ready with this PR. thanks

**Comment by Fokko** [comment]
Hi Gill, the changes are working. The only thing I have to test if we are still able to read TASK.EXTENSION-ATTEMPT using this branch.

**Comment by gilv** [comment]
@Fokko  you mean to verify we can read data that was created prior this change?

**Comment by Fokko** [comment]
Hi @gilv,

I did some tests and and I'm able to read files and directories that are written in the old format. I feel good to merge this now. If you have any further questions, please let me know.

Cheers, Fokko

**Comment by gilv** [comment]
@Fokko thanks for this patch. I will quickly review the code one more time and merge it.

**Comment by Fokko** [comment]
Hi @gilv, did you find anything? Please let me know if I can help. Cheers

**Comment by gilv** [comment]
@Fokko not yet got to it...need 1-2 days

**Comment by gilv** [comment]
@Fokko will merge it soon...sorry for delay

**Comment by Fokko** [comment]
Sure, take your time. Let me know if I can help.

**Comment by Fokko** [comment]
@gilv did you find time to look at the PR?

**Comment by Fokko** [comment]
I've rebased on master. If there are any questions that retain from merging, please let me know.

**Comment by gilv** [comment]
@Fokko thanks. i am just doing some additional tests to make 100% sure this doesn't break some existing data.. need few days

**Comment by Fokko** [comment]
Rebased onto master. @gilv did you find time to look at the PR?

**Comment by gilv** [comment]
@Fokko I did...still checking some things

**Comment by gilv** [comment]
@Fokko This patch will get into release 1.0.15 that i plan to have perhaps next week. Currently i prepare release 1.0.14

**Comment by gilv** [comment]
@Fokko can you please rebase it with current master? thanks

**Comment by Fokko** [comment]
Thanks @gilv, I've rebase onto master. If there are any further questions, please let me know.

File: src/main/java/com/ibm/stocator/fs/common/StocatorPath.java [hunk 1]
```diff
@@ -315,7 +267,11 @@ private String parseHadoopFOutputCommitterV1(Path fullPath,
             objName = fullPath.getName();
           }
           if (taskAttempt != null && !objName.startsWith(HADOOP_ATTEMPT)) {
-            objName = objName + "-" + taskAttempt;
+            // We want to prepend the attempt before the extension
+            String extension = extractExtension(objName);
+            objName = objName.replace("." + extension, "")
+                      + "-" + taskAttempt
+                      + "." + extension;
           }
```
Comment (gilv) [inline src/main/java/com/ibm/stocator/fs/common/StocatorPath.java]: @Fokko I wonder...what happens if object doesn't have extension? Do we need: replace("." + extension, "") when extension is "" and also attach "." + extension, in case there is no extension?

Comment (Fokko) [reply src/main/java/com/ibm/stocator/fs/common/StocatorPath.java]: We should write a test for this, I have time later on this week.

**Comment by Fokko** [comment]
@gilv I've added a test and it discovered a bug, thanks for the suggestion! Please let me know if I can do anything else to get this merged.

**Comment by Fokko** [comment]
@zeebonk PTAL

**Comment by Fokko** [comment]
@gilv Any further thoughts?

**Comment by zeebonk** [comment]
@Fokko @gilv this is looking great guys. I've been running the January changes since they came out. Going to run a test with changes of 5 days ago.

**Comment by gilv** [comment]
@Fokko @zeebonk I am running some tests again...

**Comment by gilv** [comment]
@Fokko I got a bit confused now..I tried to save 
myData.saveAsTextFile("swift2d://gil1903v4swift.mil01/a/data.csv.bz2")

All the objects were stored as 
a/data.csv.bz2/part-00001-attempt_20180319151051_0004_m_000001_0 

I thought your patch will add extension at the end of part-00001-attempt_20180319151051_0004_m_000001_0 ?

**Comment by Fokko** [comment]
Normally with Spark you'll write to a directory on your DFS: `swift2d://druid.mobpro/test-append/`. This is because Spark will create multiple files (equal to the partitions of the jobs). You should not define the extension yourself, this is done by Spark itself, based on the compression you define `.option("compression", "bzip2")` or `.option("compression", "snappy")`, etc. The same holds for the format, `df.write.json(..)` or `df.write.parquet(..)`, etc.

**Comment by gilv** [comment]
@Fokko I did some test and somehow hit a bit non related issue with your recent code. But the bug happens with compression only, so i decided to report it here.

import org.apache.hadoop.io.compress._
val data = Array(1, 2, 3, 4)
val myData = sc.parallelize(data)
myData.saveAsTextFile("swift2d://gil2003v2swift.mil01/a/data", classOf[GzipCodec])

Getting exception in the code you wrote before some time. If I remove classOf[GzipCodec] then all works fine.

Caused by: org.apache.spark.SparkException: Task failed while writing rows
  at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:151)
  at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:79)
  at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:78)
  at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
  at org.apache.spark.scheduler.Task.run(Task.scala:109)
  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Pipe not connected
  at java.io.PipedOutputStream.write(PipedOutputStream.java:140)
  at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
  at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
  at com.ibm.stocator.fs.swift.SwiftOutputStream.flush(SwiftOutputStream.java:240)
  at java.io.FilterOutputStream.flush(FilterOutputStream.java:140)
  at java.io.DataOutputStream.flush(DataOutputStream.java:123)
  at java.util.zip.DeflaterOutputStream.flush(DeflaterOutputStream.java:282)
  at org.apache.hadoop.io.compress.GzipCodec$GzipOutputStream.flush(GzipCodec.java:107)
  at java.io.DataOutputStream.flush(DataOutputStream.java:123)
  at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
  at org.apache.hadoop.mapred.TextOutputFormat$LineRecordWriter.close(TextOutputFormat.java:98)
  at org.apache.spark.internal.io.HadoopMapRedWriteConfigUtil.closeWriter(SparkHadoopWriter.scala:237)
  at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:133)
  at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:123)
  at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1411)
  at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:135)
  ... 8 more

**Comment by gilv** [comment]
@Fokko I also wonder about:

If i save file as swift2d://gil2003v2swift.mil01/a/data.gz then your code doesn't add .gz to the end of parts. 
If i save file as swift2d://gil2003v2swift.mil01/a/data.gz, classOf[GzipCodec]) then your code do add .gz at the end.

How you know to separate both cases? I couldn't figure out it from your code

**Comment by Fokko** [comment]
The path is rewritten my Spark internally, and the HadoopOutputFile should take care of this:
https://github.com/apache/parquet-mr/blob/master/parquet-common/src/main/java/org/apache/parquet/hadoop/metadata/CompressionCodecName.java

In your case you define the format acutally twice: `swift2d://gil2003v2swift.mil01/a/data.gz, classOf[GzipCodec])`, both by setting the extension to `.gz` and setting the `GzipCodec`.

Spark/Hadoop knows which extension to use, so you should not set this by yourself. This does not only apply to the compression, but also to the file format (`.snappy`, `.txt`, `.csv`, `.orc`, etc).

**Comment by gilv** [comment]
@Fokko  I understand..sure. But my question is different. If i write
mydata.saveAsTextFile("swift2d://gil2003v2swift.mil01/a/data.csv"), why your code doesn't catch .scv extension and adds it to the parts?

**Comment by Fokko** [comment]
Short answer, I didn't test it for corner cases like that. As you should not provide an extension by yourself in practice, so why should the code cover it? The code already too complex by itself.

**Comment by gilv** [comment]
@Fokko I guess you are right... by the way what is the diff in path pattern when extension created by the user ( as i did ) and the flow where extension is created by Spark?

**Comment by gilv** [comment]
@Fokko I know what confused me.. ignore my previous question. Your code knows to extract extensions from parts only and this what i missed. I thought the code also knows extract extensions from object names and not only parts.

**Comment by gilv** [comment]
@Fokko I was about to merge it, but then discovered some issue that i need to fix. While running bunch of tests I discovered the following:
Assume a user want to run count on Dir/*/abc/*.parquet 
User means that he wants to access all objects that satisfy this condition. With your patch parts.parquet will be returned, for example this is valid response with your patch:

Dir/*/abc/myobject.abc/part-001.-id.parquet 

This will confuse users since they don't aware of part names, only object names.. from user point of view it was not suppose to be returned, since object is myobject.abc

I know how to fix it. Will work on the fix right now

**Comment by Fokko** [comment]
@gilv Good point in testing these predicates with `*`. If I can be of any further assistance, let me know.

**Comment by gilv** [comment]
@Fokko can you please rebase it with master? thanks in advance

**Comment by Fokko** [comment]
@gilv Sure, done!

**Comment by gilv** [comment]
@Fokko very good addition. Thanks for working on this. merged

**Comment by Fokko** [comment]
Thanks @gilv for taking the time and merging. Cheers!

**Comment by zeebonk** [comment]
@gilv @Fokko Great guys, thank you very much!
